### PR TITLE
Optimise find element xi with field range caches

### LINKED ISSUE #130: Find mesh location fails to find nearest in 3D elements
The find mesh location algorithm appears unable to slide along a face or edge to get nearest, giving up early.

### LINKED ISSUE #234: Improve find mesh location performance
The find mesh location algorithm becomes very slow when meshes are large, particularly for points outside elements with find nearest, since all elements in the mesh must be searched.

Need to implement spatial sorting to be able to search through a smaller selection of elements, e.g. dividing space into an octree representation which lists all the elements which may be in each cell. As a nearest distance is found, can reduce the range of cells searched accordingly.

Requires getting accurate ranges of fields in elements, which will assist in more rapidly eliminating elements as candidates. This has other uses including visualising maximum/minimum field value point which can be added as a new element point sampling mode.

Due to the cost of creating them, find mesh location caches (ranges and octree) need to be maintained centrally to support efficient incremental graphics generation. The design must anticipate future multi-threading.

At same time should also address #130.

### PR DESCRIPTION
**Pull request by rchristie**
Fixes #234
Fixes #130
Note all find xi dimensions have some convergence issues.

### CODE REVIEW

File: src/computed_field/computed_field_find_xi_private.hpp [hunk 1]
```diff
@@ -16,81 +16,45 @@ Data structures and prototype functions needed for all find xi implementations.
...
+	void setSearchMesh(cmzn_mesh_id searchMeshIn)
 	{
-		if (new_search_mesh)
+		if (searchMeshIn)
 		{
-			cmzn_mesh_access(new_search_mesh);
+			cmzn_mesh_access(searchMeshIn);
 		}
-		if (search_mesh)
+		if (searchMesh)
```
Comment (hsorby) [inline src/computed_field/computed_field_find_xi_private.hpp]: Did you want to throw in a `this->searchMesh` in here?

File: src/computed_field/field_cache.hpp [hunk 2]
```diff
@@ -299,39 +294,43 @@ struct cmzn_fieldcache
...
 		return this->modifyCounter != this->region->getFieldModifyCounter();
 	}
 
+	/** private, only to be called by cmzn_fieldrange constructor */
+	void addFieldrange(cmzn_fieldrange *range)
+	{
+		this->fieldranges.push_back(range);
+	}
+
+	/** private, only to be called by cmzn_fieldrange denstructor */
```
Comment (hsorby) [inline src/computed_field/field_cache.hpp]: Typo `denstructor`

File: src/computed_field/field_range.hpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#if !defined (__FIELD_RANGE_HPP__)
+#define __FIELD_RANGE_HPP__
+
+#include "opencmiss/zinc/field.h"
+#include "opencmiss/zinc/fieldcache.h"
+#include "opencmiss/zinc/fieldrange.h"
+#include "computed_field/field_location.hpp"
+
+struct cmzn_fieldrange
```
Comment (hsorby) [inline src/computed_field/field_range.hpp:21]: This is meant to be a `struct`?

**Review by hsorby — CHANGES_REQUESTED** [review]

File: src/computed_field/field_range.hpp (hunk 3 above)
Comment (rchristie) [reply src/computed_field/field_range.hpp:21]: Yes, it's in the C public API, but actually I should include that api/types header to define.

**Comment by hsorby** [comment]
Other than the mentioned minors, it works for me.

**Comment by hsorby** [comment]
As a reminder, the find element xi converges well for polynomial functions but it may have convergence issues for sinusoidal functions?

**Review by hsorby — APPROVED** [review]

**Pull request merged**